src/jcpsantiago/arqivist/api/confluence/pages.clj

```diff
@@ -1,15 +1,201 @@
+   [jcpsantiago.arqivist.api.confluence.utils :as utils]
+   [jsonista.core :as json]
+   [org.httpkit.client :as http]))
+
+(defn ts->datetime
+  "Convert a UNIX timestamp into a human friendly string."
+  [ts tz]
+  (let [epoch-seconds (Long/parseLong ts)
+        zone (java.time.ZoneId/of tz)
+        formatter (java.time.format.DateTimeFormatter/ofPattern "yyyy-MM-dd HH:mm:ss")]
+    (-> epoch-seconds
+        java.time.Instant/ofEpochSecond
+        (.atZone zone)
+        (.format formatter))))
+
+;; Archival page -----------------------------------------------------------
+;; NOTE: the official Confluence markup docs are in https://confluence.atlassian.com/doc/confluence-storage-format-790796544.html
+;; TODO:
+;; * Also include message attachments
+(defn status-macro
+  "
+  Official docs in https://confluence.atlassian.com/conf59/status-macro-792499207.html
+  "
+  [title color]
+  [:ac:structured-macro {:ac:name "status"}
+   [:ac:parameter {:ac:name "title"} title]
+   [:ac:parameter {:ac:name "colour"} color]])
+
+(defn single-message-block
+  "
+  Creates the hiccup structure for a non-threaded message.
+  Will render the name of the user, the timestamp of the message,
+  and the actual text.
+  "
+  [{:keys [user-name datetime text bot?]}]
+  (list
+   (if bot?
+     [:h4 user-name " " (status-macro "bot" "Gray")
+      [:br]
+      [:code datetime]]
+     [:h4 user-name
+      [:br]
+      [:code datetime]])
+   [:p text]))
+
+(defn thread-message-block
+  "
+  Creates the hiccup structure for a threaded message.
+  The parent message is rendered on the left of the page, using `single-message-block`,
+  but the thread replies are rendered on the right. An arrow makes this distinction clearer.
+  Each reply in the thread is rendered as a normal `single-message-block`.
+  "
+  [message]
+  [:ac:layout-section {:ac:type "two_equal"
+                       :ac:breakout-mode "default"}
+   [:ac:layout-cell
+    (single-message-block message)]
+   [:ac:layout-cell
+    [:p
+     [:ac:emoticon {:ac:name "blue-star"
+                    :ac:emoji-shortname ":arrow_heading_down:"
+                    :ac:emoji-id "2935"
+                    :ac:emoji-fallback "⤵"}]
+     " " ; just a spacer, an actual <br> would be too much visually
+     (status-macro "replies" "Blue")]
+    (map single-message-block (:replies message))]])
+
+(defn page-row
+  "
+  Dispatcher fn for hiccup of different message types (single vs threaded).
+  "
+  [message]
+  (if (contains? message :replies)
+    (do
+      (mulog/log ::contains-replies
+                 :messages message)
+      (thread-message-block message))
+    (single-message-block message)))
+
+(defn card-link
+  [url]
+  (link-to
+   {:data-card-appearance "block"
+    :class "external-link"
+    :rel "nofollow"}
+   url
+   ;; confluence's API complains about XHTML parsing problems if
+   ;; what we send them has unespaced ampersands. It took me a while
+   ;; to figure this out -_-'
```

Comment:
suggest putting comment into a doc-string to make link-to function arguments easier to follow